packages/app/src/cli/prompts/dev.ts

```diff
@@ -17,17 +18,86 @@ export async function selectOrganizationPrompt(organizations: Organization[]): P
+   * as the most urgent search term, then fill in intermediate steps in the
+   * background.
+   * latestRequest tracks the most recent input, so we know which result should
+   * be displayed at all times.
+   */
+  const allInputs = ['']
+  let latestRequest: string
+
+  /* Set up an event loop, searching for apps using the latest user input once
+   * per second to avoid hitting rate limits.
+   * If we don't have any previously unseen input to search for, return without
+   * action.
+   */
+  const cachedResults: {[input: string]: OrganizationApp[]} = {'': apps}
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  const fetchInterval = setInterval(async () => {
+    let input: string | undefined
+    do {
+      input = allInputs.pop()
+      if (input === undefined) return
+    } while (cachedResults[input])
+    const result = await fetchOrgAndApps(orgId, token, input)
+    addToApiKeyCache(result.apps)
+    // eslint-disable-next-line require-atomic-updates
+    cachedResults[input] = result.apps
+  }, 1000)
+
   const choice = await ui.prompt([
     {
       type: 'autocomplete',
       name: 'apiKey',
       message: 'Which existing app is this for?',
       choices: appList,
+      /* filterFunction is a local filter-and-search, to be applied to the
+       * results from the remote search for proper sorting and display.
+       * This source function wraps the local function in a function that
+       * fetches remote results when appropriate.
+       */
+      source: (filterFunction: ui.FilterFunction): ui.FilterFunction => {
+        const cachedFiltered: {[input: string]: ui.PromptAnswer[]} = {'': appList}
+        return async (_answers: ui.PromptAnswer[], input = '') => {
+          // Only perform remote search for apps if we haven't fetched them all
+          if (appList.length < 100) return filterFunction(appList, input)
+
+          latestRequest = input
+          allInputs.push(input)
+          // Await the event loop returning a result
+          while (!cachedResults[input]) {
+            // eslint-disable-next-line no-await-in-loop
+            await system.sleep(0.2)
+          }
+
+          // Cache the answerified results to avoid race conditions
+          if (!cachedFiltered[input]) {
+            // eslint-disable-next-line require-atomic-updates
+            cachedFiltered[input] = await filterFunction(cachedResults[input]!.map(toAnswer), input)
+          }
+
+          /* If the user types "hello" we will have this function running 5
+           * times, for each substring ("h", "he", "hel", "hell", "hello").
+           * If the latest input "hello" is available, display that result.
+           * Otherwise, display the result of whatever intermediate step is
+           * available, so the user at least sees something is happening.
+           */
+          return cachedFiltered[latestRequest] || cachedFiltered[input]!
```

Comment:
I wonder wether this behavior might be a bit surprising if compared with typical web autocomplete. Could we instead display `\`no results for ${input}\``, or something similar?

Reply:
So, imagine that I type "hello" and we are loading results as we go.

If we just received results for `hell` we would show them until the results from `hello` return, and then even if the `hello` results are empty, we'd replace the `hell` results with a statement that no results were found.

So the pecking order is:

Results from the latest input, even if empty > The most recently returned results, if not empty > "Searching..."

This may matter much less anyway if we're using the debouncing strategy mentioned above.

Reply:
Got it, that makes sense 👍 And yes this will probably matter less if we go with a debouncing strategy.